crates/primitives/src/transaction/signature.rs

```diff
@@ -94,6 +94,22 @@ impl Signature {
         }
     }
 
+    /// Outputs (odd_y_parity, chain_id) from the `v` value.
+    /// This doesn't check validity of the `v` value for optimism.
+    #[inline]
+    pub fn extract_chain_id(v: u64) -> alloy_rlp::Result<(bool, Option<u64>)> {
+        if v < 35 {
+            // non-EIP-155 legacy scheme, v = 27 for even y-parity, v = 28 for odd y-parity
+            if v != 27 && v != 28 {
+                return Err(RlpError::Custom("invalid Ethereum signature (V is not 27 or 28)"))
+            }
+            Ok((v == 28, None))
+        } else {
+            // EIP-155: v = {0, 1} + CHAIN_ID * 2 + 35
+            Ok((((v - 35) % 2) != 0, Some((v - 35) >> 1)))
```

Comment:
imo this was slightly more readable
```suggestion
            // EIP-155: v = {0, 1} + CHAIN_ID * 2 + 35
            let odd_y_parity = ((v - 35) % 2) != 0;
            let chain_id = (v - 35) >> 1;
            Ok((Self { r, s, odd_y_parity }, Some(chain_id)))
```

Reply:
Why do we need `Self` on line 111?

Reply:
We don't, just took it from the prev code, just noting that we want the variable names